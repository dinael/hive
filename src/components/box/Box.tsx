import { FC, ReactNode } from 'react'
import styles from './Box.module.scss'

export type BoxProps = {
  tag?: 'div' | 'article' | 'section' | 'dialog' | 'form' | 'fieldset' | 'p' | 'blockquote' | 'picture' | 'figure' | 'code';
  space?: 'none' | 'quarter' | 'half' | 1 | 2 | 3 | 4;
  elevation?: 'xxs' | 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl' | 'pop' | 'default' | 'none';
  border?: 'none' | 'primary' | 'alt' | 'dark';
  radius?: 'none' | 'xxs' | 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl';
  paper?: boolean
  className?: string
  children?: ReactNode
}

export const Box: FC<BoxProps> = ({
  tag,
  space = 'none',
  elevation,
  border = 'primary',
  radius = 's',
  paper = false,
  className,
  children,
  ...props
}) => {

  const Tag = `${tag ? tag : 'div'}` as keyof JSX.IntrinsicElements

  const borderClass = `border:${border}`;

  const BoxStyles = `
    ${styles['box-component']}
    ${border ? styles[borderClass] : 'border:primary'}
    ${elevation ? `elevation:${elevation}` : ''}
    ${radius ? `radius:${radius}` : ''}
    ${space ? `p:${space}` : 'space:none'}
    ${paper? styles.paper : ''}
    ${className || ''}
  `

  return (
    <Tag
      className={BoxStyles}
      {...props}>
      {children}
    </Tag>
  )
}

export default Box
