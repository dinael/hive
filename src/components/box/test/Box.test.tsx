import { describe, test, expect } from 'vitest'
import { render } from '@testing-library/react'
import { Box, BoxProps } from '../Box'

describe('Box', () => {
  const defaultProps: BoxProps = {
    tag: 'div',
    space: 'none',
    elevation: 'none',
    border: 'primary',
    radius: 's',
    paper: false,
    className: 'custom-box',
    children: 'Test content',
  }

  test('Box renders snapshot', () => {
    const { container } = render(<Box {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  test('Box renders with the correct tag name', () => {
    const { container } = render(<Box {...defaultProps} />)
    const boxElement = container.querySelector('div')
    expect(boxElement?.tagName).toBe('DIV')
    expect(container).toMatchSnapshot()
  })

  test('Box renders with the correct class name', () => {
    const { container } = render(<Box {...defaultProps} />)
    const boxElement = container.querySelector('div')
    expect(boxElement?.className).toContain('custom-box')
  })

  test('Box renders with the correct border class', () => {
    const { container } = render(<Box {...defaultProps} />)
    const boxElement = container.querySelector('div')
    expect(boxElement?.className).toContain('border:primary')
  })

  test('Box renders with the correct radius class', () => {
    const { container } = render(<Box {...defaultProps} />)
    const boxElement = container.querySelector('div')
    expect(boxElement?.className).toContain('radius:s')
  })
})
