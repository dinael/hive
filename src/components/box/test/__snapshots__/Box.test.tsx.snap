// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Box > Box renders snapshot 1`] = `
<div>
  <div
    class="
    box-component
    border:primary
    elevation:none
    radius:s
    p:none
    
    custom-box
  "
  >
    Test content
  </div>
</div>
`;

exports[`Box > Box renders with the correct tag name 1`] = `
<div>
  <div
    class="
    box-component
    border:primary
    elevation:none
    radius:s
    p:none
    
    custom-box
  "
  >
    Test content
  </div>
</div>
`;
